compiler/damlc/daml-visual/src/DA/Daml/Visual.hs

```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
 addCreateChoice TemplateChoices {..} lookupData = nodeIdForChoice lookupData tplNameCreateChoice
     where tplNameCreateChoice = LF.ChoiceName $ T.pack $ DAP.renderPretty (headNote "addCreateChoice" (LF.unTypeConName (LF.tplTypeCon template))) ++ "_Create"
 
-constructSubgraphsWithLables :: Map.Map LF.ChoiceName ChoiceDetails -> TemplateChoices -> SubGraph
-constructSubgraphsWithLables lookupData tpla@TemplateChoices {..} = SubGraph nodesWithCreate template
+-- This is copied from PrettyScenarios but depending on SS for visual seems odd
+labledField :: T.Text -> T.Text -> T.Text
+labledField fname "" = fname
+labledField fname label = fname <> "." <> label
+
+typeConFieldsNames :: LF.World -> (LF.FieldName, LF.Type) -> [T.Text]
+typeConFieldsNames world (LF.FieldName fName, LF.TConApp tcn _) = map (labledField fName) (typeConFields tcn world)
+typeConFieldsNames _ (LF.FieldName fName, _) = [fName]
+
+typeConFields :: LF.Qualified LF.TypeConName -> LF.World -> [T.Text]
```

Comment:
Can this function run into an infinite loop? What happens if I have something like:

```haskell
data X = X { y : Y }
data Y = Y { x : X }
```
Admittedly that is not a useful type in DAML (since it is strict) but if you change it to something like `data X = X { y : Maybe Y }` it becomes a useful type and afaict the only reason why that won’t cause an infinite loop at the moment is that you don’t recurse into variants at all.

Reply:
That comment still applies afaict. At least it deserves a comment and maybe a TODO that we might also want to recurse on variants at some point which will require changes.

Reply:
Yes, I will add a TODO. Will revisit this again soon